presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeSourceOperator.java

```diff
@@ -103,7 +104,13 @@ public boolean isFinished()
     @Override
     public ListenableFuture<?> isBlocked()
     {
-        return source.waitForReading();
+        if (isBlocked.isDone()) {
+            isBlocked = source.waitForReading();
+            if (isBlocked.isDone()) {
```

Comment:
future of source is already cached

Reply:
Similar justification here as for `LocalExchangeSinkOperator`.